Slash Commands Permissions

msciotti opened the pull request:
This code is actually already live in production, so feel free to play around with it. Documentation is WIP not because we're not done building it, just because it might not be 100% correct yet 😄

msciotti committed:
Slash commands permissions

vladfrangu commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,17 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+> info
+> For now, disabled commands still show up in the command picker, but are unable to be used.
+
+You can also set a `default_permission` on your commands if you want them to be disabled by default when your app is added to a new guild. Setting `default_permissions` to `False` will disallow _anyone_ in a guild from using the command--even Administrators--unless a specific overwrite is configured.
```
```suggestion
You can also set a `default_permission` on your commands if you want them to be disabled by default when your app is added to a new guild. Setting `default_permissions` to `false` will disallow _anyone_ in a guild from using the command--even Administrators--unless a specific overwrite is configured.
```

vladfrangu reviewed (changes requested):
Nitpick but consistency

msciotti committed:
Permissions example

msciotti committed:
Casing

Co-authored-by: Vlad Frangu <kingdgrizzle@gmail.com>

vaporoxx commented on docs/Change_Log.md:
```diff
@@ -1,5 +1,21 @@
+New routes have been added to support this functionality:
+
+- [`GET Guild Application Command Permissions`](#DOCS_INTERACTIONS_SLASH_COMMANDS/getguildapplicationcommandpermissions)
+- [`GET Application Command Permissions`](#DOCS_INTERACTIONS_SLASH_COMMANDS/getguildapplicationcommandpermissions)
```
```suggestion
- [`GET Application Command Permissions`](#DOCS_INTERACTIONS_SLASH_COMMANDS/getapplicationcommandpermissions)
```

SpaceEEC commented on docs/interactions/Slash_Commands.md:
```diff
@@ -742,6 +782,31 @@ Edits a followup message for an Interaction. Functions the same as [Edit Webhook
 
 Deletes a followup message for an Interaction. Returns `204` on success.
 
+## Get Guild Application Command Permissions % GET /applications/{application.id#DOCS_TOPICS_OAUTH2/application-object}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/permissions
```
```suggestion
## Get Guild Application Command Permissions % GET /applications/{application.id#DOCS_TOPICS_OAUTH2/application-object}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands/permissions
```

ThaTiemsz commented on docs/Change_Log.md:
```diff
@@ -1,5 +1,21 @@
+- [`GET Application Command Permissions`](#DOCS_INTERACTIONS_SLASH_COMMANDS/getguildapplicationcommandpermissions)
+- [`PUT Application Command Permissions`](#DOCS_INTERACTIONS_SLASH_COMMANDS/putapplicationcommandpermissions)
+
+A `default_permissions` field has also been added to the [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommand) model. This field allows you to disable commands for everyone in a guild by default, if you prefer to make some of your commands an opt-in experience.
```
```suggestion
A `default_permission` field has also been added to the [ApplicationCommand](#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommand) model. This field allows you to disable commands for everyone in a guild by default, if you prefer to make some of your commands an opt-in experience.
```

ThaTiemsz commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+> info
+> For now, disabled commands still show up in the command picker, but are unable to be used.
+
+You can also set a `default_permission` on your commands if you want them to be disabled by default when your app is added to a new guild. Setting `default_permissions` to `false` will disallow _anyone_ in a guild from using the command--even Administrators--unless a specific overwrite is configured.
```
```suggestion
You can also set a `default_permission` on your commands if you want them to be disabled by default when your app is added to a new guild. Setting `default_permission` to `false` will disallow _anyone_ in a guild from using the command--even Administrators--unless a specific overwrite is configured.
```

Zomatree commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+url = "https://discord.com/api/v8/applications/<my_application_id>/guilds/<my_guild_id>/commands/<my_command_id>"
+
+json = {
+    "id": MODERATOR_ROLE_ID
```
```suggestion
    "id": MODERATOR_ROLE_ID,
```

Zomatree commented on docs/interactions/Slash_Commands.md:
````diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+
+```py
+MODERATOR_ROLE_ID = "173547401905176999"
+url = "https://discord.com/api/v8/applications/<my_application_id>/guilds/<my_guild_id>/commands/<my_command_id>"
````
```suggestion
url = "https://discord.com/api/v8/applications/<my_application_id>/guilds/<my_guild_id>/commands/<my_command_id>/permissions"
```

msciotti committed:
You are all very helpful and we appreciate you

Co-authored-by: Jan <66554238+vaporox@users.noreply.github.com>
Co-authored-by: Tiemen <ThaTiemsz@users.noreply.github.com>
Co-authored-by: Zomatree <39768508+Zomatree@users.noreply.github.com>
Co-authored-by: SpaceEEC <wermail2002-kap@yahoo.de>

whispersofthedawn commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+Need to keep some of your Slash Commands safe from prying eyes, or only available to the right people? Slash Commands support permission overwrites for all your commands. For both guild _and_ global commands, you can enable or disable a specific user or role in a guild from using a command.
+
+> info
+> For now, disabled commands still show up in the command picker, but are unable to be used.
```
It might be better to clear up that it'll be unable to be used by people who don't have the permissions, but not everyone.

whispersofthedawn commented on docs/interactions/Slash_Commands.md:
```diff
@@ -742,6 +782,31 @@ Edits a followup message for an Interaction. Functions the same as [Edit Webhook
 
 Deletes a followup message for an Interaction. Returns `204` on success.
```
Would it be worth it to specify that's it's a 204 status code? I'm not sure if this is necessary, but I wanted to give the idea.

msciotti replied:
I think consistency in our docs just has `Returns <number>`

whispersofthedawn reviewed (commented):
Looks good to me, I just had a few possible changes in mind.

ThaTiemsz reviewed (changes requested):
The JSON bodies for the POST and PATCH command endpoints (both guild and global) are missing the `default_permission` param from them:

```
| default_permission? | boolean (default `true`) | whether the command is enabled by default when the app is added to a guild |
```

Additionally, the command permissions endpoints return an object or array of objects, so a new object should be documented for them:
```
id: snowflake
application_id: snowflake
guild_id: snowflake
permissions: array of ApplicationCommandPermissions
```

GamingGeek commented on docs/interactions/Slash_Commands.md:
```diff
@@ -742,6 +782,31 @@ Edits a followup message for an Interaction. Functions the same as [Edit Webhook
+
+## Get Application Command Permissions % GET /applications/{application.id#DOCS_TOPICS_OAUTH2/application-object}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands/{command.id#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommand}/permissions
+
+Fetches command permissions for a specific command for your application in a guild. Returns an array of [ApplicationCommandPermissions](#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommandpermissions).
```
Seems these two endpoints return more than just an array of ApplicationCommandPermissions from what I can see

![image](https://user-images.githubusercontent.com/29981918/112572565-4eda8080-8de2-11eb-9ef2-08e8de13ef4d.png)

ThaTiemsz replied:
see up https://github.com/discord/discord-api-docs/pull/2737#pullrequestreview-621697678

GamingGeek replied:
> see up [#2737 (review)](https://github.com/discord/discord-api-docs/pull/2737#pullrequestreview-621697678)

Oh, didn't see that as I had files changed open while implementing it, not conversation

merlinfuchs commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+    "Authorization": "Bot 123456"
+}
+
+r = requests.put(url, headers=headers, json=json)
```
The example is wrong, the json should be:
```py
json = {
    "permissions": [
        {
             "id": MODERATOR_ROLE_ID,
             "type": 1,
             "permission": True
        }
    ]
}
```

msciotti replied:
TY ty

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+> info
+> For now, disabled commands still show up in the command picker, but are unable to be used.
+
+You can also set a `default_permission` on your commands if you want them to be disabled by default when your app is added to a new guild. Setting `default_permission` to `false` will disallow _anyone_ in a guild from using the command--even Administrators--unless a specific overwrite is configured.
```
Question because the guild owner gets weird things (like access to all channels iirc, etc.): does this impact guild owners too?

rxdn replied:
If default_permission is false, guild owners do not have access to the command

A5rocks commented on docs/interactions/Slash_Commands.md:
````diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+For example, this command will not be usable by anyone in any guilds by default:
+
+```json
+{
+    "name": "permissions_test",
+    "description": "A test of default permissions",
+    "default_permission": false
+}
````
Should this also be a python snippet like the following snippet? (for consistency.)

msciotti replied:
I think its OK since it's just the command object, and not an actual "code snippet" of something that might run

A5rocks commented on docs/interactions/Slash_Commands.md:
````diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+To enable it just for a moderator role:
+
+```py
+MODERATOR_ROLE_ID = "173547401905176999"
````
```suggestion
MODERATOR_ROLE_ID = "<moderator_role_id>"
```

Matches the following line of code by not mentioning a specific ID.

A5rocks commented on docs/interactions/Slash_Commands.md:
````diff
@@ -627,6 +627,46 @@ And, done! The JSON looks a bit complicated, but what we've ended up with is a s
+r = requests.put(url, headers=headers, json=json)
+```
+
+The command will be visible to everyone, but disabled for everyone except the moderator role.
````
Should this really document the current implementation? ("This command will be visible to everyone")

Given that just a few lines up this was already mentioned...

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -742,6 +782,31 @@ Edits a followup message for an Interaction. Functions the same as [Edit Webhook
 
+## Get Guild Application Command Permissions % GET /applications/{application.id#DOCS_TOPICS_OAUTH2/application-object}/guilds/{guild.id#DOCS_RESOURCES_GUILD/guild-object}/commands/permissions
+
+Fetches command permissions for all commands for your application in a guild. Returns an array of [ApplicationCommandPermissions](#DOCS_INTERACTIONS_SLASH_COMMANDS/applicationcommandpermissions).
```
How am I supposed to correlate the permissions and the command they belong to with this route? (if it really does only return an array of `ApplicationCommandPermissions`, which it might not... I didn't check.)

HuyaneMatsu replied:
`ApplicationCommand.id == ApplicationCommandPermissions.id`.

If an application command has no permission overwrites set in the guild, then the array wont contain an `ApplicationCommandPermissions` object for it.

A5rocks commented on docs/interactions/Slash_Commands.md:
```diff
@@ -795,6 +861,23 @@ If you specify `choices` for an option, they are the **only** valid values for a
+
+## ApplicationCommandPermissionType
+
+| Name | Value |
+| -----| ----- |
+| ROLE | 1     |
+| USER | 2     |
```
Should this be a `0` and `1` for just general consistency (see: channel overrides/overwrites/whatever they are called.)

A5rocks reviewed (commented):
Might want to run an autoformatter on the file for the tables :^)

msciotti committed:
PR review
- Clarify disabled state for commands without proper permissions
- Clarify guild owner behavior
- Fix python example
- Remove redundant disabled explanation
- Add missing default_permission on routes

msciotti committed:
default_permissions DM behavior

GitSparTV commented on docs/interactions/Slash_Commands.md:
````diff
@@ -649,13 +649,17 @@ For example, this command will not be usable by anyone in any guilds by default:
 
 ```py
-MODERATOR_ROLE_ID = "173547401905176999"
+MODERATOR_ROLE_ID = "<moderator_role_id"
````
typo

msciotti committed:
Document proper response for permissions GET

msciotti committed:
PUT all guild command permissions

msciotti committed:
typo
